### PR TITLE
fix/14490: No data displayed in options list if component using report data is rendered too early

### PR DESCRIPTION
**Pull request by jczekalski**
### Details
The issue affects components that rely on report and personal details data to display a list of options. If the component is rendered too early, the data will not be displayed even when it has fully loaded or, if personal details is not ready, the app will crash. This issue was originally identified on `SearchPage`, where, if the page is opened too early, the data would only be displayed after closing and re-opening the page or typing in the input field.

Pages/components affected by this problem:
- SearchPage
- NewChatPage / NewGroupPage
- MoneyRequestParticipantsSelector
- MoneyRequestParticipantsSplitSelector
- WorkspaceInvitePage 

The purpose of this PR is to:
- make sure the UI updates once data is available / when data changes
- display a skeleton loading state while data is loading
- prevent a crash that could occur when trying to create list options before personal details data ready
- fix crash when WorkspaceMembersPage is opened via deep link after logging out

### Fixed Issues
$ https://github.com/Expensify/App/issues/14490
$ https://github.com/Expensify/App/issues/17293
$ https://github.com/Expensify/App/issues/16739
$ https://github.com/Expensify/App/issues/18389
PROPOSAL: https://github.com/Expensify/App/issues/14490#issuecomment-1503720075

### Tests
1. Open the app
2. Set network throttling to Slow 3G or slower
3. Open one of the pages affected by the problem as soon as the app loads
- Search Page (Cmd + K)
- New Chat Page / New Group Page
- Send Money / Request Money / Split Bill -> enter amount -> tap Next
4. The options list should display a skeleton loading state
5. Data should replace the loading state once available
6. If search term is entered while data is loading, it should be applied once data is displayed

- [x] Verify that no errors appear in the JS console

#### Web
To make it easier to reproduce this issue, set network throttling to `Slow 3G` or add an even slower profile, e.g.
image

#### Mac
Simulate a bad network using Network Link Conditioner
image

### QA Steps
1. Open the app
2. Set network throttling to Slow 3G or slower
3. Open one of the pages affected by the problem as soon as the app loads
- Search Page (Cmd + K)
- New Chat Page / New Group Page
- Send Money / Request Money / Split Bill -> enter amount -> tap Next
4. The options list should display a skeleton loading state
5. Data should replace the loading state once available
6. If search term is entered while data is loading, it should be applied once data is displayed

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
Skeleton loader - Search Page (no search term)

https://user-images.githubusercontent.com/39824924/231226897-204bee35-cdae-4009-bf4e-a63ef8944df6.mov

Skeleton loader - New Chat (with search term entered while loading)

https://user-images.githubusercontent.com/39824924/231226929-8e08a7ec-d20d-4615-a1fd-cc4f86b6fa62.mov

Skeleton loader - Send Money

https://user-images.githubusercontent.com/39824924/231227022-e8e9a13e-ed33-4a68-aadb-8ea2ee582044.mov

Report and personal details data updating

https://user-images.githubusercontent.com/39824924/233602240-c8057c51-ae43-4641-8905-9f9abb49cf9e.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/39824924/231726569-f9078f6e-e5e5-4fb6-bb6f-48bc246dffeb.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/39824924/231708839-0b523568-670f-4b63-811a-0ca5eaa96250.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/39824924/231693221-3a1fde00-38c1-4902-aa00-39119c99c3ce.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/39824924/231708749-a1cee374-b841-445c-88be-e0dd5b232090.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/39824924/231726517-e9060cee-3cb1-4143-ae25-dfcb855999b9.mov

</details>

### CODE REVIEW

**Comment by jczekalski** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by jczekalski** [comment]
recheck

**Comment by jczekalski** [comment]
PR is ready for review @aimane-chnaif @amyevans

**Comment by jczekalski** [comment]
> @jczekalski don't use rebase merge. too many commits and irrelevant code changes. can you repair PR?

@aimane-chnaif Yes, sorry, I tried to update my branch with main since some conflicts showed up and I messed it up.

**Comment by jczekalski** [comment]
@aimane-chnaif Fixed

**Comment by jczekalski** [comment]
> @jczekalski please pull from main

@aimane-chnaif done

**Comment by amyevans** [comment]
@aimane-chnaif will you be able to review/test today?

**Comment by jczekalski** [comment]
> @jczekalski Did you push all the changes? I don't think PR fixes main bugs happening on search, money request pages. Showing skeleton forever, app crashes very often
> 
>  bug1.mov 
>  bug2.mov 
>  bug3.mov

@aimane-chnaif You're right, it's not working for me either. Something must have changed since I opened the PR, I'll try to figure out what's going on.

**Comment by jczekalski** [comment]
> Also let's make sure that our PR fixes #17293, #16739 which were closed in favor of our issue. So please follow their reproducible steps and fix them. The root cause and solution is well explained [here](https://github.com/Expensify/App/issues/16739#issuecomment-1498682840).

I'm not familiar with those issues, I'll take a look.

**Comment by jczekalski** [comment]
@aimane-chnaif I pulled in the changes from the latest version of `main` and everything is working correctly again.

**Comment by jczekalski** [comment]
> Let's fix #16739
> 
>  bug4.mov

@aimane-chnaif I can't reproduce this issue on my branch, even with an incredibly slow connection.

https://user-images.githubusercontent.com/39824924/233350347-05f57472-b5f5-4532-8e24-5d84ba35df60.mov

**Comment by jczekalski** [comment]
> > I can't reproduce this issue on my branch, even with an incredibly slow connection.
> 
> It's not constantly reproducible. The reason is explained [here](https://github.com/Expensify/App/issues/16739#issuecomment-1498682840). As I [stated](https://github.com/Expensify/App/issues/14490#issuecomment-1505280767) in proposal review, we should prevent calling `getNewChatOptions`, `getSearchOptions` before personalDetails data is loaded.

@aimane-chnaif Ready for review. I think all of the issues should be fixed now.

[17293](https://github.com/Expensify/App/issues/17293):

https://user-images.githubusercontent.com/39824924/233400325-a6d2933e-bb8a-4f42-bb72-54d772d77bf0.mov



[16739](https://github.com/Expensify/App/issues/16739):

https://user-images.githubusercontent.com/39824924/233400361-6596cb25-adea-40fe-afda-38cfb2947f79.mov



[This change](https://github.com/Expensify/App/pull/17349/commits/a6c58d6bcafc3978abcb21f06baf927b5bde2f22) should prevent the issue described [here](https://github.com/Expensify/App/issues/16739#issuecomment-1498682840), i.e. make sure personal details data is ready before trying to create options. Let me know what you think.

**Comment by jczekalski** [comment]
@aimane-chnaif @amyevans PR is ready for review. It should fix all of these issues: https://github.com/Expensify/App/issues/14490, 
https://github.com/Expensify/App/issues/17293, https://github.com/Expensify/App/issues/16739. I added all of them to the list of Fixed Issues in the description.

**Comment by amyevans** [comment]
@aimane-chnaif I see you added the 👀 emoji above but I'm not sure when you added it 😅. Will you be able to put this PR through its paces today?

**Comment by jczekalski** [comment]
@aimane-chnaif Any updates? Did you get a chance to review the changes?

**Review by aimane-chnaif — COMMENTED** [review]
Code looks good. So far we handled New chat/group, Search, Money request pages.
WorkspaceInvitePage also uses OptionsList so we should handle that as well.

**Comment by jczekalski** [comment]
> > [This change](https://github.com/Expensify/App/pull/17349/commits/a6c58d6bcafc3978abcb21f06baf927b5bde2f22) should prevent the issue described [#16739 (comment)](https://github.com/Expensify/App/issues/16739#issuecomment-1498682840), i.e. make sure personal details data is ready before trying to create list options. Let me know what you think.
> 
> @jczekalski I am fine with this change but just this concern: If reports data is loaded before personalDetails, what do we show? Skeleton depends on reports data loading and options list depends on personalDetails. So this case, empty list shows briefly until getting personalDetails. Most likely won't happen but we should handle this edge case as well.
> 
> @amyevans in this case, what do you suggest? still show skeleton or leave as is for now as empty list?

@aimane-chnaif @amyevans Hm, good point. Perhaps the best solution would be to add a `isLoadingPersonalDetails` Onyx key/prop in addition to `isLoadingReportData`  and display the skeleton loader until both are false?

**Comment by amyevans** [comment]
> @amyevans can you please confirm if following pages should also be fixed here?
Workspace members, Settings

Yes let's fix in this PR please.

> So if !isPersonalDetailsReady || isLoadingReportData, show skeleton

Agreed with where this landed, sorry I'm a bit late chiming in!

**Comment by jczekalski** [comment]
@aimane-chnaif I think at this point these issues are handled:
- wait for personal details and report data before hiding options list skeleton loader
- apply fixes to WorkspaceInvitePage

You also mentioned the `Workspace members` and `Settings` pages:
- `Settings`: is the issue here just adding a loading state similar to the other screens (video)?

https://user-images.githubusercontent.com/39824924/236180690-9ccd8a03-5fe5-4ba1-b3cf-ffe67d96e0e1.mov

- `Workspace members` - I couldn't reproduce the crash when opening this page, do you have any more info on this issue?

**Comment by jczekalski** [comment]
> 



> List skeleton doesn't match Settings page design. I don't see any harms to remove this code:
> 
> https://github.com/Expensify/App/blob/f5243624c029944b97d73ef8c1bcbc725f2b26d9/src/pages/settings/InitialSettingsPage.js#L277-L282
> 
> We can open wider discussion about the expected behavior on Settings page before loading personal details.

I see, but won't removing this code introduce new issues since all of the screens / components accessed from the Settings screen can now be accessed earlier than before? Perhaps it would be a good idea to move those changes to a separate issue / PR, so as not to increase the scope of this PR even more?

**Comment by jczekalski** [comment]
> > @aimane-chnaif I see, but won't removing this code introduce new issues since all of the screens / components accessed from the Settings screen can now be accessed earlier than before? Perhaps it would be a good idea to move those changes to a separate issue / PR, so as not to increase the scope of this PR even more?
> 
> I agree Settings page to be a separate issue. Thoughts @amyevans?
> 
> > all of the screens / components accessed from the Settings screen can now be accessed earlier than before?
> 
> Btw, any screens can be already accessed on web from deep link. Here's that reproducible step:
> 
> * copy deep link
> * logout
> * paste deep link into address bar and visit
> * login

Hm, good point about the deep linking. If we can move everything Settings-related to a separate issue, then perhaps this PR might be good to go? Currently it fixes this issue globally:
- prevent a crash that can occur when trying to create list options before personal details data is ready

And these issues
- make sure list options update once personal details / report data is available and whenever it changes
- display a skeleton loading state while data is loading

on these screens:
- SearchPage
- NewChatPage / NewGroupPage
- MoneyRequestParticipantsSelector
- MoneyRequestParticipantsSplitSelector
- WorkspaceInvitePage (accessed from Settings)

I updated the description of the PR to reflect the above. Waiting for your feedback @aimane-chnaif and @amyevans.

**Comment by jczekalski** [comment]
> > Workspace members - I couldn't reproduce the crash when opening this page, do you have any more info on this issue?
> 
> @jczekalski did you try this step?
> 
> * copy deep link
> * logout
> * paste deep link into address bar and visit
> * login

@aimane-chnaif Yes, the app crashes. Should I try to resolve the issue in this PR?

File: src/pages/workspace/WorkspaceInvitePage.js [hunk 1]
```diff
@@ -283,57 +286,52 @@ class WorkspaceInvitePage extends React.Component {
...
+                                isAlertVisible={this.getShouldShowAlertPrompt()}
+                                buttonText={this.props.translate('common.next')}
+                                onSubmit={this.inviteUser}
+                                message={this.props.policy.alertMessage}
+                                containerStyles={[styles.flexReset, styles.flexGrow0, styles.flexShrink0, styles.flexBasisAuto, styles.mb5]}
+                                enabledWhenOffline
+                                disablePressOnEnter
                             />
-                            <View style={[styles.flex1]}>
-                                {didScreenTransitionEnd ? (
```
[the remark this answers is not in the document]
Comment (jczekalski) [reply src/pages/workspace/WorkspaceInvitePage.js:302]: Yes, looks like I removed it accidentally along with the full screen loader. Thank you for spotting it, I added it back in. I'm trying to keep the loading state consistent with the latest version of  `SearchPage`, i.e. to replace the full screen loader

```
didScreenTransitionEnd ? <FullScreenLoader/> : <ScreenContent/>
```

with the options list skeleton
```
<OptionsSelector shouldShowOptions={didScreenTransitionEnd && otherConditions}
```

File: src/libs/OptionsListUtils.js [hunk 2]
```diff
@@ -176,6 +176,15 @@ function getPersonalDetailsForLogins(logins, personalDetails) {
     return personalDetailsForLogins;
 }
 
+/**
+ * Return true if personal details data is ready, i.e. report list options can be created.
+ * @param {Object} personalDetails
+ * @returns {boolean}
+ */
+function isPersonalDetailsReady(personalDetails) {
```
[the remark this answers is not in the document]
Comment (jczekalski) [reply src/libs/OptionsListUtils.js:179]: That was also my first thought, but I ran into a dependency cycle issue when trying to import PersonalDetailsUtils inside OptionsListUtils which I'm not sure how to resolve.
image

**Review by amyevans — DISMISSED** [review]

**Review by aimane-chnaif — DISMISSED** [review]

**Comment by jczekalski** [comment]
> There are still issues on some screens but as agreed we can fix them as separate issues. All of them can be easily tested on desktop app opened from web deep link (after logout but without re-login).
> 
> Follow-up issues:
> 
> 1. crash on split confirm (example link: http://localhost:8080/split/new/397893873612112)
> 
>  split.confirm.crash.mov 
> 2. users list not showing on split confirm (example link: http://localhost:8080/split/new/397893873612112)
> 
>  split.confirm.list.mov 
> 3. not found view briefly appears on workspace members, invite pages
> 
>  invite.mov 
>  members.mov

@aimane-chnaif Do you think we should create separate issues for these problems or perhaps one catch-all issue/PR to handle all of them? Should I create the issue(s) or should it be someone internal from Expensify?

**Comment by jczekalski** [comment]
@amyevans Conflicts resolved :)

**Review by amyevans — APPROVED** [review]

**Review by aimane-chnaif — APPROVED** [review]

**Comment by amyevans** [comment]
Okay that took a lot of re-running CI jobs due to errors: `GitHub Actions has encountered an internal error when running your job.` (https://www.githubstatus.com/), but we're passing now.

Going to merge! 🎉 I'll create the follow up issues after I get through a few other things on my plate

**Pull request merged**